helper/schema/resource_data.go

```diff
@@ -105,6 +105,14 @@ func (d *ResourceData) getRaw(key string, level getSource) getResult {
 // HasChange returns whether or not the given key has been changed.
 func (d *ResourceData) HasChange(key string) bool {
 	o, n := d.GetChange(key)
+
+	// There is a special case needed for *schema.Set's as they contain
+	// a function and reflect.DeepEqual will only say two functions are
+	// equal when they are both nil (which in this case they are not).
+	if reflect.TypeOf(o).String() == "*schema.Set" {
+		return !reflect.DeepEqual(o.(*Set).m, n.(*Set).m)
+	}
```

Comment:
LGTM but I think I'm going to use an interface for this... will CC you to show you what I mean.